File: test/python/quantifiers_test.py
from unittest import TestCase
from .parse_helper import ParseHelper
from grammars import quantifiers


class MaybeTest(TestCase, ParseHelper):
    def test_parses_a_matching_character(self):
        self.assertParse(("4", 7, []), quantifiers.parse("maybe: 4"))

    def test_parses_the_empty_string(self):
        self.assertParse(("", 7, []), quantifiers.parse("maybe: "))

    def test_rejects_a_non_matching_string(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("maybe: a")


class ZeroOrMoreTest(TestCase, ParseHelper):
    def test_parses_the_empty_string(self):
        self.assertParse(("", 7, []), quantifiers.parse("rep-0: "))

    def test_parses_one_occurrence_of_the_pattern(self):
        self.assertParse(
            ("z", 7, [
                ("z", 7, [])
            ]),
            quantifiers.parse("rep-0: z")
        )

    def test_parses_many_occurrences_of_the_same_instance_of_the_pattern(self):
        self.assertParse(
            ("zzzz", 7, [
                ("z", 7, []),
                ("z", 8, []),
                ("z", 9, []),
                ("z", 10, [])
            ]),
            quantifiers.parse("rep-0: zzzz")
        )

    def test_parses_many_occurrences_of_different_instances_of_the_pattern(self):
        self.assertParse(
            ("wxyz", 7, [
                ("w", 7, []),
                ("x", 8, []),
                ("y", 9, []),
                ("z", 10, [])
            ]),
            quantifiers.parse("rep-0: wxyz")
        )

    def test_rejects_strings_with_a_non_matching_prefix(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-0: 4x")

    def test_rejects_strings_with_a_non_matching_suffix(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-0: x4")

    def test_parses_repeating_patterns_greedily(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("greedy-0: xy")


class OneOrMoreTest(TestCase, ParseHelper):
    def test_rejects_the_empty_string(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-1: ")

    def test_parses_one_occurrence_of_the_pattern(self):
        self.assertParse(
            ("z", 7, [
                ("z", 7, [])
            ]),
            quantifiers.parse("rep-1: z")
        )

    def test_parses_many_occurrences_of_the_same_instance_of_the_pattern(self):
        self.assertParse(
            ("zzzz", 7, [
                ("z", 7, []),
                ("z", 8, []),
                ("z", 9, []),
                ("z", 10, [])
            ]),
            quantifiers.parse("rep-1: zzzz")
        )

    def test_parses_many_occurrences_of_different_instances_of_the_pattern(self):
        self.assertParse(
            ("wxyz", 7, [
                ("w", 7, []),
                ("x", 8, []),
                ("y", 9, []),
                ("z", 10, [])
            ]),
            quantifiers.parse("rep-1: wxyz")
        )

    def test_rejects_strings_with_a_non_matching_prefix(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-1: 4x")

    def test_rejects_strings_with_a_non_matching_suffix(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-1: x4")

    def test_parses_repeating_patterns_greedily(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("greedy-1: xy")

    def test_parses_a_repeated_reference(self):
        self.assertParse(
            ("#abc123", 11, [
                ("#", 11, []),
                ("abc123", 12, [
                    ("a", 12, []),
                    ("b", 13, []),
                    ("c", 14, []),
                    ("1", 15, []),
                    ("2", 16, []),
                    ("3", 17, [])
                ])
            ]),
            quantifiers.parse("color-ref: #abc123")
        )

    def test_parses_a_repeated_choice(self):
        self.assertParse(
            ("#abc123", 14, [
                ("#", 14, []),
                ("abc123", 15, [
                    ("a", 15, []),
                    ("b", 16, []),
                    ("c", 17, []),
                    ("1", 18, []),
                    ("2", 19, []),
                    ("3", 20, [])
                ])
            ]),
            quantifiers.parse("color-choice: #abc123")
        )


class ExactlyTest(TestCase, ParseHelper):
    def test_rejects_the_empty_string(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-exact: ")

    def test_parses_the_required_number_of_the_pattern(self):
        self.assertParse(
            ("abc", 11, [
                ("a", 11, []),
                ("b", 12, []),
                ("c", 13, [])
            ]),
            quantifiers.parse("rep-exact: abc")
        )

    def test_rejects_too_few_copies_of_the_pattern(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-exact: ab")

    def test_rejects_too_many_copies_of_the_pattern(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-exact: abcd")


class MinimumTest(TestCase, ParseHelper):
    def test_rejects_the_empty_string(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-min: ")

    def test_parses_the_minimum_number_of_the_pattern(self):
        self.assertParse(
            ("abc", 9, [
                ("a", 9, []),
                ("b", 10, []),
                ("c", 11, [])
            ]),
            quantifiers.parse("rep-min: abc")
        )

    def test_parses_more_copies_of_the_pattern(self):
        self.assertParse(
            ("abcdef", 9, [
                ("a", 9, []),
                ("b", 10, []),
                ("c", 11, []),
                ("d", 12, []),
                ("e", 13, []),
                ("f", 14, [])
            ]),
            quantifiers.parse("rep-min: abcdef")
        )

    def test_rejects_too_few_copies_of_the_pattern(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-min: ab")


class RangeTest(TestCase, ParseHelper):
    def test_rejects_the_empty_string(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-range: ")

    def test_parses_the_minimum_number_of_the_pattern(self):
        self.assertParse(
            ("abc", 11, [
                ("a", 11, []),
                ("b", 12, []),
                ("c", 13, [])
            ]),
            quantifiers.parse("rep-range: abc")
        )

    def test_parses_the_maximum_number_of_the_pattern(self):
        self.assertParse(
            ("abcde", 11, [
                ("a", 11, []),
                ("b", 12, []),
                ("c", 13, []),
                ("d", 14, []),
                ("e", 15, [])
            ]),
            quantifiers.parse("rep-range: abcde")
        )

    def test_rejects_too_few_copies_of_the_pattern(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-range: ab")

    def test_rejects_too_many_copies_of_the_pattern(self):
        with self.assertRaises(quantifiers.ParseError):
            quantifiers.parse("rep-range: abcdef")

File: test/java/QuantifiersTest.java
package canopy.quantifiers;

import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertThrows;
import org.junit.jupiter.api.Test;

import helpers.Node;
import helpers.NodeSpec;

import test.grammars.quantifiers.Label;
import test.grammars.quantifiers.ParseError;
import test.grammars.quantifiers.Quantifiers;
import test.grammars.quantifiers.TreeNode;

class MaybeTest extends ParseHelper {
    @Test
    void parsesAMatchingCharacter() throws ParseError {
        expect(Quantifiers.parse("maybe: 4")).toMatch(node("4", 7).noElems());
    }

    @Test
    void parsesTheEmptyString() throws ParseError {
        expect(Quantifiers.parse("maybe: ")).toMatch(node("", 7).noElems());
    }

    @Test
    void rejectsANonMatchingCharacter() {
        assertThrows(ParseError.class, () -> Quantifiers.parse("maybe: a"));
    }
}

class ZeroOrMoreTest extends ParseHelper {
    @Test
    void parsesTheEmptyString() throws ParseError {
        expect(Quantifiers.parse("rep-0: ")).toMatch(node("", 7).noElems());
    }

    @Test
    void parsesOneOccurrenceOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-0: z")).toMatch(
            node("z", 7)
                .elem(node("z", 7).noElems())
        );
    }

    @Test
    void parsesManyOccurrencesOfTheSameInstanceOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-0: zzzz")).toMatch(
            node("zzzz", 7)
                .elem(node("z", 7).noElems())
                .elem(node("z", 8).noElems())
                .elem(node("z", 9).noElems())
                .elem(node("z", 10).noElems())
        );
    }

    @Test
    void parsesManyOccurrencesOfDifferentInstancesOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-0: wxyz")).toMatch(
            node("wxyz", 7)
                .elem(node("w", 7).noElems())
                .elem(node("x", 8).noElems())
                .elem(node("y", 9).noElems())
                .elem(node("z", 10).noElems())
        );
    }

    @Test
    void rejectsStringsWithANonMatchingPrefix() {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-0: 4x"));
    }

    @Test
    void rejectsStringsWithANonMatchingSuffix() {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-0: x4"));
    }

    @Test
    void parsesRepeatingPatternsGreedily() {
        assertThrows(ParseError.class, () -> Quantifiers.parse("greedy-0: xy"));
    }
}

class OneOrMoreTest extends ParseHelper {
    @Test
    void rejectsTheEmptyString() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-1: "));
    }

    @Test
    void parsesOneOccurrenceOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-1: z")).toMatch(
            node("z", 7)
                .elem(node("z", 7).noElems())
        );
    }

    @Test
    void parsesManyOccurrencesOfTheSameInstanceOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-1: zzzz")).toMatch(
            node("zzzz", 7)
                .elem(node("z", 7).noElems())
                .elem(node("z", 8).noElems())
                .elem(node("z", 9).noElems())
                .elem(node("z", 10).noElems())
        );
    }

    @Test
    void parsesManyOccurrencesOfDifferentInstancesOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-1: wxyz")).toMatch(
            node("wxyz", 7)
                .elem(node("w", 7).noElems())
                .elem(node("x", 8).noElems())
                .elem(node("y", 9).noElems())
                .elem(node("z", 10).noElems())
        );
    }

    @Test
    void rejectsStringsWithANonMatchingPrefix() {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-1: 4x"));
    }

    @Test
    void rejectsStringsWithANonMatchingSuffix() {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-1: x4"));
    }

    @Test
    void parsesRepeatingPatternsGreedily() {
        assertThrows(ParseError.class, () -> Quantifiers.parse("greedy-1: xy"));
    }

    @Test
    void parsesARepeatedReference() throws ParseError {
        expect(Quantifiers.parse("color-ref: #abc123")).toMatch(
            node("#abc123", 11)
                .elem(node("#", 11).noElems())
                .elem(node("abc123", 12)
                    .elem(node("a", 12).noElems())
                    .elem(node("b", 13).noElems())
                    .elem(node("c", 14).noElems())
                    .elem(node("1", 15).noElems())
                    .elem(node("2", 16).noElems())
                    .elem(node("3", 17).noElems())
                )
        );
    }

    @Test
    void parsesARepeatedChoice() throws ParseError {
        expect(Quantifiers.parse("color-choice: #abc123")).toMatch(
            node("#abc123", 14)
                .elem(node("#", 14).noElems())
                .elem(node("abc123", 15)
                    .elem(node("a", 15).noElems())
                    .elem(node("b", 16).noElems())
                    .elem(node("c", 17).noElems())
                    .elem(node("1", 18).noElems())
                    .elem(node("2", 19).noElems())
                    .elem(node("3", 20).noElems())
                )
        );
    }
}

class ExactlyTest extends ParseHelper {
    @Test
    void rejectsTheEmptyString() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-exact: "));
    }

    @Test
    void parsesTheRequiredNumberOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-exact: abc")).toMatch(
            node("abc", 11)
                .elem(node("a", 11).noElems())
                .elem(node("b", 12).noElems())
                .elem(node("c", 13).noElems())
        );
    }

    @Test
    void rejectsTooFewCopiesOfThePattern() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-exact: ab"));
    }

    @Test
    void rejectsTooManyCopiesOfThePattern() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-exact: abcd"));
    }
}

class MinimumTest extends ParseHelper {
    @Test
    void rejectsTheEmptyString() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-min: "));
    }

    @Test
    void parsesTheRequiredNumberOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-min: abc")).toMatch(
            node("abc", 9)
                .elem(node("a", 9).noElems())
                .elem(node("b", 10).noElems())
                .elem(node("c", 11).noElems())
        );
    }

    @Test
    void parsesMoreCopiesOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-min: abcdef")).toMatch(
            node("abcdef", 9)
                .elem(node("a", 9).noElems())
                .elem(node("b", 10).noElems())
                .elem(node("c", 11).noElems())
                .elem(node("d", 12).noElems())
                .elem(node("e", 13).noElems())
                .elem(node("f", 14).noElems())
        );
    }

    @Test
    void rejectsTooFewCopiesOfThePattern() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-min: ab"));
    }
}

class RangeTest extends ParseHelper {
    @Test
    void rejectsTheEmptyString() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-range: "));
    }

    @Test
    void parsesTheMinimumNumberOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-range: abc")).toMatch(
            node("abc", 11)
                .elem(node("a", 11).noElems())
                .elem(node("b", 12).noElems())
                .elem(node("c", 13).noElems())
        );
    }

    @Test
    void parsesTheMaximumNumberOfThePattern() throws ParseError {
        expect(Quantifiers.parse("rep-range: abcde")).toMatch(
            node("abcde", 11)
                .elem(node("a", 11).noElems())
                .elem(node("b", 12).noElems())
                .elem(node("c", 13).noElems())
                .elem(node("d", 14).noElems())
                .elem(node("e", 15).noElems())
        );
    }

    @Test
    void rejectsTooFewCopiesOfThePattern() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-range: ab"));
    }

    @Test
    void rejectsTooManyCopiesOfThePattern() throws ParseError {
        assertThrows(ParseError.class, () -> Quantifiers.parse("rep-range: abcdef"));
    }
}

class ParseHelper {
    Node<Label> expect(TreeNode node) {
        return new NodeWrapper(node.elements.get(1));
    }

    NodeSpec<Label> node(String text, int offset) {
        return new NodeSpec<Label>(text, offset);
    }
}

class NodeWrapper implements Node<Label> {
    private TreeNode node;

    NodeWrapper(TreeNode node) {
        this.node = node;
    }

    public String text() {
        return node.text;
    }

    public int offset() {
        return node.offset;
    }

    public List<Node<Label>> elements() {
        return node.elements.stream()
            .map((elem) -> new NodeWrapper(elem))
            .collect(Collectors.toList());
    }

    public Node<Label> get(Label key) {
        return new NodeWrapper(node.get(key));
    }
}
